compiler/phpdoc.cpp

```diff
@@ -21,115 +21,195 @@
 #include "compiler/utils/string-utils.h"
 #include "compiler/ffi/ffi_parser.h"
 
-using std::vector;
-using std::string;
-
-const std::map<string, php_doc_tag::doc_type> php_doc_tag::str2doc_type = {
-  {"@param",                     param},
-  {"@kphp-inline",               kphp_inline},
-  {"@kphp-flatten",              kphp_flatten},
-  {"@kphp-infer",                kphp_infer},
-  {"@kphp-required",             kphp_required},
-  {"@kphp-lib-export",           kphp_lib_export},
-  {"@kphp-sync",                 kphp_sync},
-  {"@kphp-should-not-throw",     kphp_should_not_throw},
-  {"@kphp-throws",               kphp_throws},
-  {"@type",                      var},
-  {"@var",                       var},
-  {"@return",                    returns},
-  {"@returns",                   returns},
-  {"@kphp-disable-warnings",     kphp_disable_warnings},
-  {"@kphp-extern-func-info",     kphp_extern_func_info},
-  {"@kphp-pure-function",        kphp_pure_function},
-  {"@kphp-template",             kphp_template},
-  {"@kphp-param",                kphp_param},
-  {"@kphp-return",               kphp_return},
-  {"@kphp-memcache-class",       kphp_memcache_class},
-  {"@kphp-immutable-class",      kphp_immutable_class},
-  {"@kphp-tl-class",             kphp_tl_class},
-  {"@kphp-const",                kphp_const},
-  {"@kphp-no-return",            kphp_noreturn},
-  {"@kphp-warn-unused-result",   kphp_warn_unused_result},
-  {"@kphp-warn-performance",     kphp_warn_performance},
-  {"@kphp-analyze-performance",  kphp_analyze_performance},
-  {"@kphp-serializable",         kphp_serializable},
-  {"@kphp-reserved-fields",      kphp_reserved_fields},
-  {"@kphp-serialized-field",     kphp_serialized_field},
-  {"@kphp-serialized-float32",   kphp_serialized_float32},
-  {"@kphp-profile",              kphp_profile},
-  {"@kphp-profile-allow-inline", kphp_profile_allow_inline},
-  {"@kphp-strict-types-enable",  kphp_strict_types_enable},
-  {"@kphp-color",                kphp_color},
-};
+struct KnownPhpDocTag {
+  unsigned int hash;
+  PhpDocType type;
+  const char *tag_name;
 
-vector<php_doc_tag> parse_php_doc(vk::string_view phpdoc) {
-  if (phpdoc.empty()) {
-    return {};
+  constexpr KnownPhpDocTag(const char *tag_name, PhpDocType type)
+    : hash(5381), type(type), tag_name(tag_name) {
+    for (const char *c = tag_name; *c; ++c) {
+      hash = (hash << 5) + hash + *c;
+    }
   }
+};
 
-  int line_num_of_function_declaration = stage::get_line();
+class AllDocTags {
+  static constexpr int N_TAGS = 35;
+  static const KnownPhpDocTag ALL_TAGS[N_TAGS];
 
-  vector<string> lines(1);
-  bool have_star = false;
-  for (char c : phpdoc) {
-    if (!have_star) {
-      if (c == ' ' || c == '\t') {
-        continue;
+public:
+  static PhpDocType name2type(const char *start, const char *end) __attribute__((always_inline)) {
```

Comment:
Please use c++17 attributes in new code. Here it will be [[gnu::always_inline]] prior function definition.

Reply:
Done. Didn't know about that!